bridges/snowbridge/primitives/router/src/inbound/v2.rs

```diff
@@ -0,0 +1,642 @@
+/// in a beneficiary account.
+#[derive(Clone, Encode, Decode, RuntimeDebug, TypeInfo)]
+pub enum Asset {
+	NativeTokenERC20 {
+		/// The native token ID
+		token_id: H160,
+		/// The monetary value of the asset
+		value: u128,
+	},
+	ForeignTokenERC20 {
+		/// The foreign token ID
+		token_id: H256,
+		/// The monetary value of the asset
+		value: u128,
+	},
+}
+
+/// Reason why a message conversion failed.
+#[derive(Copy, Clone, TypeInfo, PalletError, Encode, Decode, RuntimeDebug, PartialEq)]
+pub enum ConvertMessageError {
+	/// Invalid foreign ERC-20 token ID
+	InvalidAsset,
+	/// Cannot reachor a foreign ERC-20 asset location.
+	CannotReanchor,
+}
+
+pub trait ConvertMessage {
+	fn convert(
+		message: Message,
+		origin_account: Location,
+	) -> Result<(Xcm<()>, u128), ConvertMessageError>;
+}
+
+pub struct MessageToXcm<
+	EthereumNetwork,
+	InboundQueuePalletInstance,
+	ConvertAssetId,
+	GatewayProxyAddress,
+	EthereumUniversalLocation,
+	GlobalAssetHubLocation,
+> where
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueuePalletInstance: Get<u8>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	GatewayProxyAddress: Get<H160>,
+	EthereumUniversalLocation: Get<InteriorLocation>,
+	GlobalAssetHubLocation: Get<Location>,
+{
+	_phantom: PhantomData<(
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	)>,
+}
+
+impl<
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	> ConvertMessage
+	for MessageToXcm<
+		EthereumNetwork,
+		InboundQueuePalletInstance,
+		ConvertAssetId,
+		GatewayProxyAddress,
+		EthereumUniversalLocation,
+		GlobalAssetHubLocation,
+	>
+where
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueuePalletInstance: Get<u8>,
+	ConvertAssetId: MaybeEquivalence<TokenId, Location>,
+	GatewayProxyAddress: Get<H160>,
+	EthereumUniversalLocation: Get<InteriorLocation>,
+	GlobalAssetHubLocation: Get<Location>,
+{
+	fn convert(
+		message: Message,
+		origin_account_location: Location,
+	) -> Result<(Xcm<()>, u128), ConvertMessageError> {
+		let mut message_xcm: Xcm<()> = Xcm::new();
+		if message.xcm.len() > 0 {
+			let xcm_string = hex::encode(message.xcm.clone());
+			log::info!(target: LOG_TARGET,"found xcm payload: {:x?}", xcm_string);
+			// Allow xcm decode failure so that assets can be trapped on AH instead of this
+			// message failing but funds are already locked on Ethereum.
+			if let Ok(versioned_xcm) = VersionedXcm::<()>::decode_with_depth_limit(
+				MAX_XCM_DECODE_DEPTH,
+				&mut message.xcm.as_ref(),
+			) {
+				if let Ok(decoded_xcm) = versioned_xcm.try_into() {
+					message_xcm = decoded_xcm;
+				} else {
+					log::error!(target: LOG_TARGET,"unable to decode xcm");
+				}
+			} else {
+				log::error!(target: LOG_TARGET,"unable to decode versioned xcm");
+			}
```

Comment:
failing this should fail the whole conversion, no?

short-form suggestion:
```suggestion
			if let Ok(decoded_xcm) = VersionedXcm::<()>::decode_with_depth_limit(
				MAX_XCM_DECODE_DEPTH,
				&mut message.xcm.as_ref(),
			).map(|versioned_xcm| versioned_xcm.try_into()) {
				message_xcm = decoded_xcm;
			} else {
				let xcm_string = hex::encode(message.xcm);
				log::debug!(target: LOG_TARGET, "unable to decode xcm: {:x?}", xcm_string);
				return Err(ConvertMessageError::InvalidXcm);
			}
```

Reply:
@acatangiu should we not rather fail silent on invalid xcm, so that assets can be trapped on AH? Otherwise the funds will be locked on Ethereum but not claimable on Polkadot. This was @alistair-singh 's idea.